Clarify variability of functions. (#2924)
modelica/ModelicaSpecification: chapters/interface.tex

  constraining interface of the function being redeclared.
\end{itemize}

> Note that function-compatibilty does currently not ensure that the function call has the correct variability, and that must then be checked after re-declaration - due the possibility of default arguments with higher variability.

Reviewer: En-dash and spelling:
```suggestion
Note that function-compatibility does currently not ensure that the function call has the correct variability, and that must then be checked after re-declaration -- due the possibility of default arguments with higher variability.
```

Reviewer 2: Also missing "to":
```suggestion
Note that function-compatibility does currently not ensure that the function call has the correct variability, and that must then be checked after re-declaration -- due to the possibility of default arguments with higher variability.
```

> Note that function-compatibility does currently not ensure that the function call has the correct variability, and that must then be checked after re-declaration -- due to the possibility of default arguments with higher variability.

Reviewer: This is an important note, but I think we should aim for a situation where we don't need to say that our variability definitions are wrong.  How about something like this (combined with additional changes in _Variability of Expressions_ to be suggested next):
```suggestion
Note that variability of function calls cannot be determined using the just the interface of a function, as the variabilities of default argument expressions are not expressed by the interface.
Hence a function redeclaration being function-compatible does not ensure that function calls will fulfill variability requirements, and tools must therefore check variability requirements separately.

\begin{nonnormative}
One way that tools may carry out the additional checks is to first infer the variability requirements on functions that can be redeclared, but looking at the expressions where the function is called.
The interface of the function that could be redeclared can then be extended with a variability requirement for any default argument expressions, allowing errors to be detected and reported at the point of a redeclaration that introduces too high variability through default argument expressions.
\end{nonnormative}
```

By the way, we only have _re-declaration_ written with a hyphen in one other place, and we should probably get rid of the hyphen also in that case.

Reviewer 3: there is a the too much here

Author: Most of this was merged, but the non-normative part reduced; so I think this is resolved.

Reviewer: > there is a the too much here

Fixed.

Reviewer: > Most of this was merged, but the non-normative part reduced; so I think this is resolved.

To me, the non-normative part looks stripped rather than reduced.  I'd like to leave this conversation open, as the non-normative part is closely related to the ongoing discussion below in https://github.com/modelica/ModelicaSpecification/pull/2924#discussion_r642131665.

Author: I think we can handle that later, and merge this first as it is difficult to keep track of everything.

\begin{nonnormative}
Thus any call of a function that is replaceable will need to be checked after redeclarations.
\end{nonnormative}

> 

Reviewer: This looks like unnecessary repetition.  Can it simply be dropped?

Author: I dropped the non-normative texts that seemed redunddant and added a reference in the other part.

Reviewer: I'm sure exactly what added reference you mean, but the unnecessary repetition seems fixed.  Resolving.

  constraining interface of the function being redeclared.
\end{itemize}

> Note that variability of function calls cannot be determined using the just the interface of a function, as the variabilities of default argument expressions are not expressed by the interface.

Reviewer 3: using the just -> using just

Reviewer: You're suggesting this:
```suggestion
Note that variability of function calls cannot be determined using just the interface of a function, as the variabilities of default argument expressions are not expressed by the interface.
```

  constraining interface of the function being redeclared.
\end{itemize}

> Note that function-compatibilty does currently not ensure that the function call has the correct variability, and that must then be checked after re-declaration.

Reviewer: I find this kind of formulation quite controversial.  If we know our concepts are broken in this way, I think it makes more sense to have an open issue than to describe the limitation normatively.  If we should go the other way, the least we should do before adding a formulation like this – in my opinion – is to collect informed comments from all tool vendors.

Author: To me it makes sense to formulate the current limitations and later work on it separately - as explained above that it is not trivial - especially if we want to preserve compatibility for MSL.
However, if it helps I can remove that.

Author: The question is basically: how does tools handle models such as Modelica.Fluid.Machines.BaseClasses.PartialPump with a function-default having a parameter-value to ensure that the function call isn't used for a constant in the original model?

Since I haven't seen any previous complaints from any tool vendors about this construct I assume tools didn't do anything special and thus there is a "hole" in the check as stated in the new statement.

Author: @gkurzbach @sjoelund @eshmoylova consider Modelica.Fluid.Machines.ControlledPump and Modelica.Fluid.Machines.BaseClasses.PartialPump

And especially adding `final constant .Modelica.Units.SI.Position dzero = flowCharacteristic(0);` to PartialPump. 

Would you detect an error in PartialPump?
An error in the redeclare in the extends-clause of ControlledPump:
```
  redeclare replaceable function flowCharacteristic =
        Modelica.Fluid.Machines.BaseClasses.PumpCharacteristics.quadraticFlow (
          V_flow_nominal={0, V_flow_op, 1.5*V_flow_op},
          head_nominal={2*head_op, head_op, 0}))
```
Or some other special logic to detect this?
Alternatively is it just handled as part of normal variability checks (as suggested by this edit)?

Reviewer 4: OpenModelica uses the normal variability checks and gives the following error:
```
Error: Component dzero of variability constant has binding 'HeatingSystem_total.pump.flowCharacteristic(0)' of higher variability parameter.
```

Author: > OpenModelica uses the normal variability checks and gives the following error:
> 
> ```
> Error: Component dzero of variability constant has binding 'flowCharacteristic(0)' of higher variability parameter.
> ```

And in case someone wonders: Dymola does the same (with a similar error message).

Reviewer: > Since I haven't seen any previous complaints from any tool vendors about this construct I assume tools didn't do anything special and thus there is a "hole" in the check as stated in the new statement.

I can understand if tools are forced to open up holes because of shortcomings in the specification, but I'd like to avoid making this the prescribed way, as it could make it more difficult to develop a more solid variability logic in the future.

Reviewer: > OpenModelica uses the normal variability checks and gives the following error:

Note that this discussion is about calling this _normal_ or not.  To me, it's not normal; the normal analysis would be that we have a function call with only constant arguments, hence it's a constant expression.  In other words, having parameter variability hiding in the (even replaceable!) function `flowCharacteristic` is _not_ normal procedure according to my reading of the current text.

Author: > > OpenModelica uses the normal variability checks and gives the following error:
> 
> Note that this discussion is about calling this _normal_ or not. 

But the proposed text doesn't say that it is _normal_, only that it hasn't been checked:
"Note that function-compatibilty does currently not ensure that the function call has the correct variability, and that must then be checked after re-declaration."

That just states that we _currently_ have missing guarantee here, and thus you need this extra check, so I don't see the problem with this text as a solution until we possibly find a better solution. I can, of course, remove this text or make it non-normative - but I don't see why we would be hung up on a word that isn't present in the text.

Reviewer: The problem is that it is a normative statement telling that the way we define variability of function calls in 3.8 _Variability of Expressions_ is incorrect.

Author: > The problem is that it is a normative statement telling that the way we define variability of function calls in 3.8 _Variability of Expressions_ is incorrect.

No, it's saying that it's incomplete  - as these hidden arguments are also considered.
It could be clarified to say:
"Note that function-compatibilty does currently not ensure that the function call has the correct variability, and that must then be checked after re-declaration - by also considering the variability of these added arguments."

Reviewer 2: In MapleSim we do the variability check on the arguments of `flowCharacteristic(0)` and do not check any any default arguments a function may have. So we do not detect the error for `dzero` in `ControlledPump`.

Reviewer 4: In OpenModelica we first construct a function call with a complete argument list as specified in [12.4.1](https://specification.modelica.org/master/functions.html#positional-or-named-input-arguments-of-functions) and then we type check and determine the variability of the call. But I guess it's not specified that this is the way it must be done.

Author: I agree that there is a problem, but as I see it currently there are two possibilities - either you variability-check with the extra argument (as indicated by the new text; and already done by Dymola and OpenModelica) - or you check it earlier and thus you end up with equations breaking the variability-guarantees.

Obviously it would be better if this problem didn't exist - but until we have solved it I find it best to have it documented.

>   end UseDriveLine;

Reviewer: I'm not afraid of resolving merge conflicts, so I propose that you make a separate PR for getting rid of `UseDriveline`.  OK?

>   end UseDriveLine;

Reviewer 2: I am having difficulties creating a PR, so I created an issue #2968.

>   end UseDriveLine;

Author: That separate issue is now closed.

  end UseDriveLine;
  Modelica.Mechanics.MultiBody.Interface.Frame_a frame_a;
  replaceable function gravity = GravityInterface;

>   constant Real failed[:]=gravity({1,0,0}); // May fail

Reviewer: Spacing:
```suggestion
  constant Real failed[:] = gravity({1, 0, 0}); // May fail
```

equation
  frame_a.f = gravity(frame_a.r0);
  // or gravity(position=frame_a.r0);
  frame_a.t = zeros(3);
end Body;

model PlanetSimulation

>   parameter Modelica.Units.SI.Mass mSun=2e30;
>   function sunGravity = PointMassGravity (m=mSun);

Reviewer: Spacing:
```suggestion
  parameter Modelica.Units.SI.Mass mSun = 2e30;
  function sunGravity = PointMassGravity (m = mSun);
```
